trie/smt.py

```diff
@@ -0,0 +1,355 @@
+                             key in the keypath
+
+        """
+        validate_is_bytes(key)
+        validate_length(key, self._key_size)
+
+        # Path diff is the logical XOR of the updated key and this account
+        path_diff = (to_int(self.key) ^ to_int(key))
+
+        # Same key (diff of 0), update the tracked value
+        if path_diff == 0:
+            self._value = value
+            # No need to update branch
+        else:
+            # Find the first mismatched bit between keypaths. This is
+            # where the branch point occurs, and we should update the
+            # sibling node in the source branch at the branch point.
+            # NOTE: Keys are in MSB->LSB (root->leaf) order.
+            #       Node lists are in root->leaf order.
+            #       Be sure to convert between them effectively.
+            for bit in reversed(range(self._branch_size)):
+                if path_diff & (1 << bit) > 0:
+                    branch_point = (self._branch_size-1)-bit
+                    break
+
+            # NOTE: node_updates only has to be as long as necessary
+            #       to obtain the update. This allows an optimization
+            #       of pruning updates to the maximum possible depth
+            #       that would be required to merge, which may be many
+            #       orders of magnitude smaller than the tree depth.
+            if len(node_updates) <= branch_point:
+                raise ValidationError("Updated node list is not deep enough")
+
+            # Update sibling in the branch where our key differs from the update
+            self._branch[branch_point] = node_updates[branch_point]
+            # No need to update value
+
+
+class SparseMerkleTree:
+
+    def __init__(self, key_size: int=32, default: bytes=BLANK_NODE):
+        # Ensure we can support the given depth
+        if not 1 <= key_size <= 32:
+            raise ValidationError("Keysize must be number of bytes in range [0, 32]")
+
+        self._key_size = key_size  # key size in bytes
+        self.depth = key_size * 8  # depth is number of bits in key
+
+        self._default = default
+
+        # Initialize an empty tree with one branch
+        self.db = {}
+        node = self._default  # Default leaf node
+        for _ in range(self.depth):
+            node_hash = keccak(node)
+            self.db[node_hash] = node
+            node = node_hash + node_hash
+
+        # Finally, write the root hash
+        self.root_hash = keccak(node)
+        self.db[self.root_hash] = node
+
+    @classmethod
+    def from_db(
+            cls,
+            db: Dict[bytes, bytes],
+            root_hash: Hash32,
+            key_size: int=32,
+            default: bytes=BLANK_NODE):
+        smt = cls(key_size=key_size, default=default)
+
+        # If db is provided, and is not consistent,
+        # there may be a silent error. Can't solve that easily.
+        for key, value in db.items():
```

Comment:
> Can you point to an example?

Say, all of the state in Ethereum mainnet. Even though plyvel technically supports key iteration it would be prohibitively expensive computationally.

> Also, what do you think about this being a separate class method?

I think if this becomes like the other tries, with a reference to the underlying db, then most of this method can disappear, and the rest moves to init.

Reply:
Well, unlike the Ethereum tries, there are two characteristics that I think warrant the separate approach.
1. SMTs can be used in a more ad-hoc fashion.  For example in Plasma, each transaction in a block is bundled into an SMT and summarized via the root, which is uploaded to Ethereum. Probably not too different from the txn root in Ethereum, but wortth noting.
2. You may want to specify a different default value for an empty database to the trie, so that it can be used for different characteristics. For example, storage in Ethereum assumes a default value for the key, which is `bytes32(0)` instead of `b''`. I was trying to allow for this behavior, and disambiguate it from db storage use cases through the separate class method.

Reply:
On 1) nothing stops you from passing in a {} as the "underlying database", that's a totally reasonable use case that py-evm does sometimes, too.

On 2) if we want to be careful about it, `__init__` could always try to add the default values to the DB (maybe checking if they exist first). Other than that, it seems like we could still follow the same pattern for the trie init.

Reply:
Re: 1) {} would not be a valid database, because it does not have a pathway to the default value.
Re: 2) because of 1), not sure if mixing this logic is the best choice because that would have an eiher/or constructor layout

Reply:
1) huh? `self.db = {}` already in [the PR](https://github.com/ethereum/py-trie/pull/77/files/4f2e5f5e66f553b96bda5b76281a12309b2f16b3#diff-730d445fc351b9302eb71dc655598097R205)
2) I'm also cool with having a separate `smt = SMT.init_db(db, default_value=...)`, where `SMT(db, default_value=...)` would assume the database is already initialized.

Reply:
1. Immediately followed by creating the branch for the default value in the db through the iteration lol
2. Cool, thanks. I think this is clearer that there are two use cases.

Reply:
Solved the original concern by just directly assigning the db in the classmethod.